analytic_engine/src/table_meta_set_impl.rs

```diff
@@ -0,0 +1,326 @@
+// Copyright 2023 CeresDB Project Authors. Licensed under Apache-2.0.
+
+//! Table data set impl based on spaces
+
+use std::{fmt, sync::Arc};
+
+use common_util::error::BoxError;
+use log::debug;
+use snafu::{OptionExt, ResultExt};
+use table_engine::table::TableId;
+
+use crate::{
+    manifest::{
+        details::{
+            ApplySnapshotToTableWithCause, ApplyUpdateToTableNoCause, ApplyUpdateToTableWithCause,
+            BuildSnapshotNoCause, TableMetaSet,
+        },
+        meta_edit::{
+            self, AddTableMeta, AlterOptionsMeta, AlterSchemaMeta, DropTableMeta, MetaEditRequest,
+            MetaUpdate, VersionEditMeta,
+        },
+        meta_snapshot::MetaSnapshot,
+    },
+    space::{Space, SpaceId, SpacesRef},
+    sst::file::FilePurgerRef,
+    table::{
+        data::{TableData, TableShardInfo},
+        version::{TableVersionMeta, TableVersionSnapshot},
+        version_edit::VersionEdit,
+    },
+};
+
+#[derive(Clone)]
+pub(crate) struct TableMetaSetImpl {
+    pub(crate) spaces: SpacesRef,
+    pub(crate) file_purger: FilePurgerRef,
+    // TODO: maybe not suitable to place this parameter here?
+    pub(crate) preflush_write_buffer_size_ratio: f32,
+}
+
+impl fmt::Debug for TableMetaSetImpl {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("spaces table snapshot provider")
+    }
+}
+
+impl TableMetaSetImpl {
+    fn apply_update(
+        &self,
+        meta_update: MetaUpdate,
+        shard_info: TableShardInfo,
+    ) -> crate::manifest::details::Result<()> {
+        // `FnOnce` can only be pass as the instance but not reference, however `Impl
+        // FnOnce` can not act as the parameter of the closure... So Box<dyn
+        // FnOnce> is used here.
+        let find_space_and_apply_edit = |space_id: SpaceId,
+                                         apply_edit: Box<
+            dyn FnOnce(Arc<Space>) -> crate::manifest::details::Result<()>,
+        >| {
+            let spaces = self.spaces.read().unwrap();
+            let space = spaces
+                .get_by_id(space_id)
+                .with_context(|| ApplyUpdateToTableNoCause {
+                    msg: format!("space not found, space_id:{space_id}"),
+                })?;
+            apply_edit(space.clone())
+        };
```

Comment:
Why not define this as a method on the `TableMetaSetImpl` with apply_edit as a generic parameter?